llama.cpp

```diff
@@ -8146,7 +8185,15 @@ struct llm_build_context {
                 cb(cur, "ffn_out", il);
             }
 
-            cur = ggml_add(ctx0, cur, ffn_inp);
+            if (model.layers[il].ffn_norm) {
+                // non-parallel residual
+                cur = ggml_add(ctx0, cur, ffn_inp);
+            } else {
+                // add together residual + FFN + self-attention
+                cur = ggml_add(ctx0, cur, inpL);
+                cur = ggml_add(ctx0, cur, attn_out);
+            }
```

Comment:
Aren't these 2 branches equivalent?

Reply:
I don't believe so. One is doing parallel residual (eg 12b) and the other when the ffn norm is present (eg stablelm 1.6 and 3b) is not doing parallel residual. If I am missing something please let me know thanks !

Reply:
Since `ffn_inp = attn_out + inpL` I think these branches do the same and can be replaced by simply with:

```c++
cur = ggml_add(ctx0, cur, ffn_inp);
```

I am looking for ways to avoid the unused `ffn_inp = ggml_add(...)` in the parallel-residual case

Reply:
> Since `ffn_inp = attn_out + inpL` I think these branches do the same

Reasoning from [the relevant modeling code in `transformers`](https://github.com/huggingface/transformers/blob/487505ff45b1bd134a210853e1339cb205ba77b3/src/transformers/models/stablelm/modeling_stablelm.py#L769-L782), even though they separate them for clarity, you're right, theses branches are the same.

Reply:
@ggerganov Thanks! I removed the branches. Please let me know if there is anything else !